add method to capture NSError

ahti opened the pull request:
This PR introduces a method to capture NSErrors. It's a WIP and I'd appreciate any feedback. It also doesn't include any tests yet.

### Message content

`raven-swift` and `raven-objc` have similar methods. This implementation, unlike theirs, does not send the errors description as message, as I find that to have bad readability and it may exceed the limit of 1000 characters for the message.

Instead, it sends messages similar to this:

    NSURLErrorDomain.-1004 in SomeFile.swift:99 someFunc(_:params:)

I'd like to send a more readable error name, but the only thing available at runtime is `localizedDescription`, which can be in any number of languages.

A possibility to do this would be to ship a dictionary that maps error domains and codes to their respective names, but generating such a dictionary automatically is not trivial and maintaining it manually seems unfeasible (there are at least 57 error domains in the iOS 9.3 SDK).

### userInfo

In addition, the `userInfo` dictionary is cleaned to be convertible to JSON and sent in `extra["user_info"]`.

I'm not a fan of how I implemented this and I'd love suggestions on how it could be done better. I've tried to integrate this into the `EventSerializable` protocol, but couldn't make it work due to the infamous `Protocol 'EventSerializable' can only be used as a generic constraint because it has Self or associated type requirements. `.

### Source location

I added `#file`, `#line` and `#function` parameters to the capture function. We might want to extend this to other functions and also send a single-frame stacktrace (this is what the other frameworks do), as the culprit doesn't seem to be displayed anywhere.

I have also not yet tested how (if at all) this is usable from Objective-C.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/getsentry/sentry-swift/12)
<!-- Reviewable:end -->

joshdholtz commented:
@ahti I will definitely try to take a look at this later this afternoon is possibly otherwise tomorrow morning. (I just moved over the weekend as I mentioned in another PR so trying to catch up to all the things I missed 😉  ).

Thanks for submitting what you have so far! I really appreciate the contribution and I will get back to you by tomorrow with any thoughts 👌

ahti commented:
I rebased onto master and added tests etc. I also factored out the `SourceLocation` into an internal class that can be used in other `capture...`-messages as well (I can adapt `captureMessage` in a separate PR after this is merged).

I'm still not a friend of `cleanValue` and friends, but can't find a better solution either.

In the future, I'd love to find a solution to sending more than the numeric error code, but for now I think this is good enough (and better than what the other iOS clients do).

dcvz commented:
I think it may be best to leave it up to the developer to do this since it allows the developer to format the `NSError` or their own `ErrorType` as they wish. (i.e.)
```swift
switch type {
case .Error(let error):
    message = "\(error.domain)\n\(error.localizedDescription)"
case .AppError(let error):
    message = "\(error.domain).\(error.identifier)"
}
...
let event = Sentry.Event.build(message) {
    $0.level = .Error
    $0.culprit = "\(function) in \((file as NSString).lastPathComponent)"
    $0.releaseVersion = Device.currentDevice.appVersion
}
client?.captureEvent(event)
```
Given the Sentry seperates events based on the message, this should allow you to get unique entries for every error. It would be great if we somehow had control over setting the description line as well (or if it could take it from the second line of our message). It would also be nice if we could set the culprit to show on the event though :( https://github.com/getsentry/sentry/issues/3308 because then we don't really need to add one line of stack trace.

Given also that we have breadcrumbs it should give enough information that led up to the error. With all those things it probably wouldn't be necessary to have this added.

joshdholtz commented:
Sentry does not separate based on message/name - events are grouped by stacktrace but can be overridden by fingerprints.

That being said, I still feel like what this PR is adding will bring some value to this client. One change that might be nice is to change the message that is being sent.

Instead of.. 
```
NSURLErrorDomain.-1004 in SomeFile.swift:99 someFunc(_:params:)
```
... the message could be... 
```
NSURLErrorDomain (1004) - Could not connect to the server.
```
...and then use the `sentry.interfaces.stacktrace.Stacktrace` interface (https://docs.getsentry.com/hosted/clientdev/interfaces/) which will show the stacktrace nicely in the UI (and group by that stacktrace)?

Let me know your thoughts when you get a chance!  Just woke up and didn't have enough ☕  yet so let me know if that doesn't make sense 😇

dcvz commented:
About how to label them.. after messing around with and filing this issue: https://github.com/getsentry/sentry/pull/3323. I finally was able to produce this result using the `exceptions` interface:
<img width="1232" alt="screen shot 2016-05-25 at 18 11 32" src="https://cloud.githubusercontent.com/assets/2475932/15547754/80f70a28-22a5-11e6-933b-dfe75751e9d6.png">

I'm now adding the interface to the SDK so then maybe we could use that for this :) I'm setting the title as `error.domain` and the description as `error.localizedDescription`

ahti commented:
Using the exceptions interface seems like a smart idea!

The problem with using the `localizedDescription` is that it's localized. If a user is using an iPhone set to e.g. german language, an error might look something like this (NSURLErrorDomain.-1005): "Die Netzwerkverbindung wurde unterbrochen." So a single error might come in under multiple names, and most of those names would not be readable for the dev.

dcvz commented:
Hey, you're in Germany too! I'm in Berlin :)

That's actually a great point, I just tested it and indeed it separates the events. Maybe it's best to set the title as NSError and the description as `error.domain (error.code)` and setting the userInfo in the extras. The issues will be separated as they should and you'll be able to gather more details from the culprit and extras.

I've added basic exceptions support [here](https://github.com/getsentry/sentry-swift/pull/16/commits/87c39f83ecce4ebd9a09b28afb4b8d1ed91356d8) (without stacktrace), so hopefully we'll see that in soon.

dcvz commented:
I played around with different options:
<img width="1247" alt="screen shot 2016-05-26 at 08 33 30" src="https://cloud.githubusercontent.com/assets/2475932/15565697/a63cbfe0-231c-11e6-9429-331836b56adf.png">
vs
<img width="1241" alt="screen shot 2016-05-26 at 08 32 04" src="https://cloud.githubusercontent.com/assets/2475932/15565699/ab22da44-231c-11e6-893d-7a0a89a32771.png">

I think the second options looks better and makes it easier for devs to find things quickly

ahti commented:
The second one looks great, I think!

So if you folks agree I guess we can put this PR on ice until your exception-implementation is mergerd and then I'll rebase and adapt it to use that interface?

dcvz commented:
@ahti the PR adding the `Exception` interface has been merged in c59210eece9d301f5b419e5ad184e4fc91e35de9. You can go ahead and incorporate that into the error reporting 👍 

<img width="603" alt="screen shot 2016-05-31 at 17 06 40" src="https://cloud.githubusercontent.com/assets/2475932/15679436/1e1a9e0e-2752-11e6-9625-b217161d2228.png">

This is what I do in my project maybe it can lead some direction :) where `errorValue` is `"\(error.domain) (\(error.code))"`

ahti commented:
Sorry for taking so long with this.

I added the exception interface, rebased onto master and fixed some small things after the rebase.

While I'm still not 100% satisfied with sending error codes, I have not yet found a way to get better messages. So I think until we find one, this will be good enough.

Do you need any more changes before this can be merged?

ahti commented:
@joshdholtz do you need any more changes before this can be merged?

joshdholtz commented:
@ahti It looks like there is a merge conflict still. Can you resolve that? 😊

ahti commented:
@joshdholtz I rebased onto current master, tests are running fine for me, so this should be good to merge now

joshdholtz commented:
Thanks @ahti! Good work on this 👍

joshdholtz commented on Sources/Sentry+Error.swift:
```diff
@@ -0,0 +1,76 @@
+
+extension SentryClient {
+
+    // broken out into a separate function for testability
```
❤️
